core/common/src/main/scala/org/typelevel/otel4s/meta/InstrumentMeta.scala

```diff
@@ -17,45 +17,104 @@
 
-  /** Modify the context `F` using an implicit [[KindTransformer]] from `F` to `G`.
-    */
-  def mapK[G[_]](implicit kt: KindTransformer[F, G]): InstrumentMeta[G] =
-    mapK(kt.liftK)
-}
+    /** Runs `f` only when the instrument is enabled. A shortcut for `Monad[F].ifM(isEnabled)(f, unit)`.
+      */
+    def whenEnabled(f: => F[Unit]): F[Unit]
 
-object InstrumentMeta {
+    /** Modify the context `F` using the transformation `f`. */
+    def mapK[G[_]: Monad](f: F ~> G): InstrumentMeta.Dynamic[G] =
+      new Dynamic.MappedK(this)(f)
+
+    /** Modify the context `F` using an implicit [[KindTransformer]] from `F` to `G`.
+      */
+    def mapK[G[_]: Monad](implicit kt: KindTransformer[F, G]): InstrumentMeta.Dynamic[G] =
+      mapK(kt.liftK)
+  }
 
-  def enabled[F[_]: Applicative]: InstrumentMeta[F] =
-    make(enabled = true)
+  object Dynamic {
 
-  def disabled[F[_]: Applicative]: InstrumentMeta[F] =
-    make(enabled = false)
+    private[otel4s] def from[F[_]: Monad](enabled: F[Boolean]): Dynamic[F] =
+      new Dynamic[F] {
+        def isEnabled: F[Boolean] = enabled
+        def unit: F[Unit] = Monad[F].unit
+        def whenEnabled(f: => F[Unit]): F[Unit] = Monad[F].ifM(isEnabled)(f, unit)
+      }
 
-  private def make[F[_]: Applicative](enabled: Boolean): InstrumentMeta[F] =
-    new InstrumentMeta[F] {
-      val isEnabled: Boolean = enabled
-      val unit: F[Unit] = Applicative[F].unit
+    private[otel4s] def enabled[F[_]: Applicative]: Dynamic[F] =
+      new Dynamic[F] {
+        val isEnabled: F[Boolean] = Applicative[F].pure(true)
+        val unit: F[Unit] = Applicative[F].unit
+        def whenEnabled(f: => F[Unit]): F[Unit] = f
+      }
+
+    private[otel4s] def disabled[F[_]: Applicative]: Dynamic[F] =
+      new Dynamic[F] {
+        val isEnabled: F[Boolean] = Applicative[F].pure(false)
+        val unit: F[Unit] = Applicative[F].unit
+        def whenEnabled(f: => F[Unit]): F[Unit] = unit
+      }
+
+    private class MappedK[F[_], G[_]: Monad](meta: InstrumentMeta.Dynamic[F])(f: F ~> G) extends Dynamic[G] {
+      def isEnabled: G[Boolean] = f(meta.isEnabled)
+      def unit: G[Unit] = f(meta.unit)
+      def whenEnabled(f: => G[Unit]): G[Unit] = Monad[G].ifM(isEnabled)(f, Monad[G].unit)
     }
+  }
+
+  sealed trait Static[F[_]] {
+
+    /** Indicates whether instrumentation is enabled or not.
+      */
+    def isEnabled: Boolean
 
-  private class MappedK[F[_], G[_]](meta: InstrumentMeta[F])(f: F ~> G) extends InstrumentMeta[G] {
-    def isEnabled: Boolean = meta.isEnabled
-    def unit: G[Unit] = f(meta.unit)
+    /** A no-op effect.
+      */
+    def unit: F[Unit]
+
+    /** Modify the context `F` using the transformation `f`. */
+    def mapK[G[_]](f: F ~> G): InstrumentMeta.Static[G] =
+      new Static.MappedK(this)(f)
+
+    /** Modify the context `F` using an implicit [[KindTransformer]] from `F` to `G`.
+      */
+    def mapK[G[_]](implicit kt: KindTransformer[F, G]): InstrumentMeta.Static[G] =
```

Comment:
completely orthogonal to this PR, but I think all the `mapK` methods that take a `KindTransformer` should be called `liftTo` (or something similar). just something to think about for after the migration of KT to mtl

Reply:
I wish *so* dearly. but no. we've at least reached a point where I think Ross and I are in consensus? but we still have to at a minimum decide which members of which typeclass hierarchy make it in, sorry =/